### ISSUE TITLE
Literal types are converted to str when combining dictionaries with syntactic sugars

### ISSUE THREAD

**Issue by TheodoreWolf** [microsoft/pyright#6474]
**Describe the bug**
When combining dictionaries with syntactic sugars where the key is a `Literal`, the type of the key is converted to `str`. For example: `dict[Literal[str], T] | dict[Literal[str], T]` should have the type `dict[Literal[str], T]`, but gets` dict[str,T]`.

**Code or Screenshots**

```python
from typing import Literal

Animal = Literal["dog", "cat"]

def combine_with_syntactic_sugar_1(
    c: dict[Animal, int], d: dict[Animal, int]
) -> dict[Animal, int]:
    return c | d # line 8


def combine_with_syntactic_sugar_2(
    c: dict[Animal, int], d: dict[Animal, int]
) -> dict[Animal, int]:
    return {**c, **d} # line 14


def combine_no_sugar(c: dict[Animal, int], d: dict[Animal, int]) -> dict[Animal, int]:
    _dict = {}
    for k, v in c.items():
        _dict[k] = v
    for k, v in d.items():
        _dict[k] = v
    return _dict
```
Where we get the following errors: 
```sh
pyright_bug.py:8:12 - error: Expression of type "dict[str, int]" cannot be assigned to return type "dict[Animal, int]"
    "dict[str, int]" is incompatible with "dict[Animal, int]"
      TypeVar "_KT@dict" is invariant
        Type "str" cannot be assigned to type "Animal"
          "str" cannot be assigned to type "Literal['dog']" (reportGeneralTypeIssues)
pyright_bug.py:14:12 - error: Expression of type "dict[str, int]" cannot be assigned to return type "dict[Animal, int]" (reportGeneralTypeIssues)
```
We can see from the line numbers that `combine_no_sugar()` passes typechecking but the two other syntactic sugar functions fail (with slightly different errors).

python 3.9.18
pyright 1.1.294

**Comment by erictraut** [comment]
You are using a very old version of pyright (42 revisions ago!). Please make sure you're using the latest version before you report a bug. The latest version is 1.1.336.

I'm not able to repro the issue with the latest version, so I'm closing the issue.
